ios/amfi/amfi.go

```diff
@@ -0,0 +1,174 @@
+	}
+
+	if _, ok := plist["success"]; ok {
+		return nil
+	}
+
+	return fmt.Errorf("EnableDevMode: could not enable developer mode through amfi service but no error or success was reported")
+}
+
+// When you enable developer mode and device is rebooted, you get a popup on the device to finish enabling developer mode
+// This function "accepts" that popup
+func (devModeConn *Connection) EnableDevModePostRestart() error {
+	reader := devModeConn.deviceConn.Reader()
+
+	request := map[string]interface{}{"action": 2}
+
+	bytes, err := devModeConn.plistCodec.Encode(request)
+	if err != nil {
+		return fmt.Errorf("EnableDevModePostRestart: failed encoding request to service with err: %w", err)
+	}
+
+	err = devModeConn.deviceConn.Send(bytes)
+	if err != nil {
+		return fmt.Errorf("EnableDevModePostRestart: failed sending request bytes to service with err: %w", err)
+	}
+
+	responseBytes, err := devModeConn.plistCodec.Decode(reader)
+	if err != nil {
+		return fmt.Errorf("EnableDevModePostRestart: failed decoding response from service with err: %w", err)
+	}
+
+	plist, err := ios.ParsePlist(responseBytes)
+	if err != nil {
+		return fmt.Errorf("EnableDevModePostRestart: failed parsing response plist with err: %w", err)
+	}
+
+	if _, ok := plist["success"]; ok {
+		return nil
+	}
+
+	return fmt.Errorf("EnableDevModePostRestart: could not enable developer mode post restart through amfi service")
+}
+
+func EnableDeveloperMode(device ios.DeviceEntry, enablePostRestart bool) error {
+	// Don't try to enable if it already is
+	devModeEnabled, err := imagemounter.IsDevModeEnabled(device)
+	if err != nil {
+		return fmt.Errorf("EnableDeveloperMode: failed checking developer mode status with err: %w", err)
+	}
+
+	if devModeEnabled {
+		log.Info("Developer mode is already enabled on the device")
+		return nil
+	}
+
+	// Perform the first step of developer mode enablement and wait for the device to restart
+	conn, err := New(device)
+	if err != nil {
+		return fmt.Errorf("EnableDeveloperMode: failed connecting to amfi service with err: %w", err)
+	}
+
+	err = conn.EnableDevMode()
+	if err != nil {
+		return fmt.Errorf("EnableDeveloperMode: failed enabling developer mode with err: %w", err)
+	}
+	log.Infof("Successfully enabled developer mode on device `%s`, device will restart", device.Properties.SerialNumber)
+
+	udid := device.Properties.SerialNumber
+	log.Info("Waiting for device to restart after enabling developer mode")
+	ticker := time.NewTicker(10 * time.Second)
+	defer ticker.Stop()
+
+	// Loop trying to reinit the device to find out if it restarted
+WaitLoop:
+	for {
+		select {
+		case <-ticker.C:
+			device, err = ios.GetDevice(udid)
+			if err != nil {
+				log.Info("Device is not yet available")
+				continue WaitLoop
+			}
+			break WaitLoop
+		case <-time.After(60 * time.Second):
+			ticker.Stop()
+			if err != nil {
+				return errors.New("Device was not restarted in 60 seconds")
+			}
```

Comment:
I think the nicer solution would be to use a `contextContext` that gets passed to this function instead of hardcoding this into the implementation.

Also, this parts is anyways never reached. The ticker fires every 10s, and on the next loop iteration we create a new `time.After` channel (and that's another problem to look out in general 😉 `time.After` always creates a new goroutine, which can't be stopped if we don't need it anymore. Meaning those resources are always only released after the timeout exceeds)

Reply:
Yeah, now that I think about it you are actually right, this does reset the timer on each iteration. Using context sounds like a reasonable solution, I might update the code these days. Initially this wrapping function was only in the CLI code and I didn't think much of it when you could just kill it if it hangs, shame on me :D